docs/design/proposals/VRG-Reconciliation.md

```diff
@@ -0,0 +1,132 @@
+# VRG reconciler workflow
+
+This design document covers the VRG CRD and its reconciliation workflow.
+
+**NOTE**: This is added to aid understanding and ease development of the
+various parts in parallel (where possible) and would possibly be merged once
+it reaches initial completion.
+
+## Terminology
+
+- VRG: VolumeReplicationGroup
+- VR: VolumeReplication
+- PVC: PersistentVolumeClaim
+- PV: PersistentVolume
+
+## VRG CRD
+
+- Spec:
+    - PVCSelector
+        - Label selector to identify all the PVCs that are in this group that
+        needs to be replicated to the peer cluster.
+    - VolumeReplicationClass
+        - ReplicationClass of all volumes in this replication group; this
+        value is propagated to children VolumeReplication CRs
+    - ReplicationState
+        - Desired state of all volumes [`primary` or `secondary`] in this
+        replication group; this value is propagated to children
+        VolumeReplication CRs
+    - S3Endpoint
+        - S3 Endpoint to replicate PV metadata
+    - S3SecretName
+        - S3 Endpoint to replicate PV metadata
+- Status:
+    - **TBD**
+
+### VRG Example
+
+```
+---
+apiVersion: ramendr.openshift.io/v1alpha1
+kind: VolumeReplicationGroup
+metadata:
+  name: vrg-sample
+spec:
+  pvcSelector:
+    matchLabels:
+      appname: myapp
+  volumeReplicationClass: "vr-class-sample"
+  replicationState: "Primary"
+  s3Endpoint: http://objectstore.svc:9000
+  s3SecretName: vrg-objectstore-secret
+```
+
+## Reconciling VRG as Primary
+
+1. Select all PVCs that match `PVCSelector` for processing
+1. For each PVC
+    1. If not bound, continue but requeue
+    1. Check `vr-protected` finalizer, if found,
+        1. Check and remove (stale) `vr-protecting` finalizer
+        1. Check and update VR as `primary` and continue
+    1. Add `vr-protecting` finalizer to PVC, to prevent PVC deletion till VRG,
+    and hence VR is deleted
+    1. Change PV `reclaimPolicy` to "Retain", if not already "Retain" and add
+    an annotation to remember the change
+    1. [Find|Create] VR for PVC as `primary`
+    1. Backup PV bound to PVC
+    1. Add `vr-protected` finalizer to PVC
+    1. Remove `vr-protecting` finalizer
+1. For each PVC
+    1. Check if VR status has reached Primary state
+        1. If not, update status as progressing
+    1. Report status as completed, if not progressing for any PVC
+
+### Notes
+
+- `vr-protecting` and `vr-protected` are provided as 2 finalizers, to aid in
+  optimizing reconciliation if a PVC is already `vr-protected`
```

Comment:
Added some clarity on the optimization.

Reply:
I read the latest version: "vr-protecting and vr-protected are provided as 2 finalizers, to aid in optimizing reconciliation if a PVC is already vr-protected".  Perhaps I'm having trouble with the sentence because of recursive use of vr-protected.  I think my question is that is it merely an optimization or a requirement to do it 2 phases to correctly handle crashes?  Are you trying to use this to atomically do the sequence of retain policy, create VolRep, and upload PV?  

For instance, if we do the following then we may not need the two phases. For each PVC, find VolRep.  If VolRep is not found, change retain policy, upload PV and then create VolRep as primary.  In this case, the two steps prior to creating VolRep is idempotent and hence, doesn't require two phases.  This is why I'm asking if you're doing the two phases for a reason (for correctness) or if it just an optimization.

Reply:
Went through the 2 finalizer case again, which I **thought** I picked the pattern from the snapshot extrenal controller.

The short of it is, we only need this to optimize calls to add finalizer to PVC, backup PV, change PV retention policy (as of now).

This should not be 2 finalizers, rather it should be a single finalizer and possibly an added annotation that states PVC protection is complete, for the required optimization.

With that, we would add a single finalizer to protect the PVC from getting deleted when operations are in flight, and also to ensure we can set VR state as required before PVC deletion.

The added annotation can come in to optimize out other processing as stated, and can be added towards the end of processing the PVC and VR.

Now, in delete and moving to Secondary state, we expect the PVC to be deleted, so we cannot update the annotation (IOW create it at the end of processing), these cases would require special handling to not assume protected state and complete all the *required* steps of reconciliation.

Reply:
LGTM